src/Microsoft.DotNet.Interactive/Parsing/DirectiveParser.cs

```diff
@@ -0,0 +1,59 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.CommandLine;
+using System.CommandLine.Builder;
+using System.CommandLine.Parsing;
+
+namespace Microsoft.DotNet.Interactive.Parsing
+{
+    internal class DirectiveParser
+    {
+        private readonly Parser _commandLineParser;
+        public DirectiveParser(RootCommand rootCommand)
+        {
+            var commandLineBuilder =
+                    new CommandLineBuilder(rootCommand)
+                        .ParseResponseFileAs(ResponseFileHandling.Disabled)
+                        .UseTypoCorrections()
+                        .UseHelpBuilder(bc => new DirectiveHelpBuilder(rootCommand.Name))
+                        .UseHelp()
+                        .UseMiddleware(
+                            context =>
+                            {
+                                context.BindingContext
+                                       .AddService(
+                                           typeof(KernelInvocationContext),
+                                           _ => KernelInvocationContext.Current);
+                            });
+            commandLineBuilder.EnableDirectives = false;
+            _commandLineParser = commandLineBuilder.Build();
+        }
+
+        public ParseResult Parse(string input)
+        {
+            if (FindComment(input) is { } realPos)
+                input = input.Substring(0, realPos);
+            return _commandLineParser.Parse(input);
+        }
+
+        private static int? FindComment(string input)
+        {
+            var firstQuote = input.IndexOf('"');
+            if (firstQuote is -1)
+                return null;
+
+            var secondQuote = input.IndexOf('"', firstQuote + 1);
+            if (secondQuote is -1)
+                return null;
+
+            var doubleSlash = input.IndexOf("//", secondQuote + 1);
+            if (doubleSlash is -1)
+                return null;
+
+            return doubleSlash;
+        }
```

Comment:
It would be useful to start with a few test cases, possibly using `[Theory]`. Quote-counting might not be valuable, since it's a) not required for `#r`, e.g. `#r nuget:package` is valid, and b) it's valid to surround any token with quotes, and it's commonly used to indicate a single token when there's a space in it, e.g. `#!blah "this is parsed as one token"`.

As for test cases, none of the following should be identified as containing a comment:

```console
#!value --from-file //volume/path/to/file
#!value --from-url https://example.com
```

Reply:
I thought they aren't directives. If they are, the thing becomes way less trivial. Here's what I see as possible solutions:
- Make a wrapper around `ParseResult` too. The reason is that after we got a `ParseResult` from the command line parser, we need to see if the unmatched tokens are in fact comments. Since `ParseResult` is immutable (which is, in fact, good), we need to extract the necessary data from it, draw our own conclusions regarding comments, and then return our `DirectiveParseResult`.
- Change the behaviour. Make `#r nuget:DuckPackage //quack` raise an error as with invalid syntax (now it swallows errors with no feedback).

What do you think?

Reply:
Magic commands and directives are the same thing at the implementation level. If there's any semantic distinction it's that a "directive" refers to something that the compiler is or someday will be aware of, i.e. `#i` and `#r`. (Support for these is currently not at parity across C# and F#.)

Since a comment like this is valid for a directive in the underlying language, I think it's useful to fix this. A more targeted fix, just for `#i` and `#r`, might be easier. Distinguishing a comment from valid inputs above will still be challenging though.

Reply:
Given that the command line parser can distinguish, I can reinterpret its parse result and see, if the first unmatched token starts with `//`. Okay, let me see if I actually can. I will prepare a number of sets, and you will say if you think that's the behaviour you'd expect.